### PR TITLE
Adds a check in `sample` for `start` kwarg shapes

### PR DESCRIPTION
**Pull request by jackhansom**
Addresses #2323 

This PR is designed to check the the shapes of arguments specified in the `start` kwargs are valid given the model `vars`. I ran the tests locally and they all passed. 
Main things missing at this point are:
- potential update to `sample` docstring
- unit tests

Something I realised while writing this was that a list of dicts is a valid argument for `start` (as it is used in test_parallel_start.py), but this is not documented in the docstring and I would need some pointers as to what the accepted types are (if there are any beyond dict and list of dict) to update the docstring properly.

If whoever looks at this is generally happy with the approach, I can also add some unit tests (I was thinking in test_sampling.py)

### CODE REVIEW

File: pymc3/sampling.py [hunk 1]
```diff
@@ -280,6 +283,37 @@ def sample(draws=500, step=None, init='auto', n_init=200000, start=None,
     return sample_func(**sample_args)[discard:]
 
 
+def _check_start_shape(model, start):
+    e = ''
+    for var in model.vars:
+        if np.shape(start):
```
Comment (aseyboldt) [inline pymc3/sampling.py]: Shouldn't this be outside of the loop over the vars?
The check if `np.shape()` is a bit strange. How about using `isinstance(start, (collections.abc.Sequence, np.ndarray))` instead? I didn't know that `np.shape` returns an empty tuple before I checked in the interpreter, and I think most people wouldn't.

File: pymc3/sampling.py [hunk 2]
```diff
@@ -280,6 +283,37 @@ def sample(draws=500, step=None, init='auto', n_init=200000, start=None,
...
+            # to deal with iterable start argument
+            for start_iter in start:
+                _check_start_shape(model, start_iter)
+        elif not isinstance(start, dict):
+            raise TypeError("start argument must be a dict"
```
Comment (aseyboldt) [inline pymc3/sampling.py]: missing space between string literals

**Review by aseyboldt — COMMENTED** [review]
Nice. A test with invalid input for `sample` using `pytest.raises` would be nice.

File: pymc3/sampling.py (hunk 1 above)
Comment (jackhansom) [reply pymc3/sampling.py]: Yes, that might be more clear. Basically I saw that this was how `_make_parallel` checked to see if `arg` was array-like, so I used the same expression. But I'm happy to change it to something more clear.

Comment (jackhansom) [reply pymc3/sampling.py]: And yes, that check should be placed outside the loop over vars

File: pymc3/sampling.py [hunk 3]
```diff
@@ -280,6 +283,37 @@ def sample(draws=500, step=None, init='auto', n_init=200000, start=None,
     return sample_func(**sample_args)[discard:]
 
 
+def _check_start_shape(model, start):
+    e = ''
+    if isinstance(start, (Sequence, np.ndarray)):
+        # to deal with iterable start argument
+        for start_iter in start:
+            _check_start_shape(model, start_iter)
```
Comment (aseyboldt) [inline pymc3/sampling.py:291]: missing `return`

**Comment by aseyboldt** [comment]
A test would still be nice. You can use `pytest.raises` to check exceptions.

File: pymc3/sampling.py (hunk 3 above)
Comment (jackhansom) [reply pymc3/sampling.py:291]: Added into latest commits

**Comment by aseyboldt** [comment]
The test failure is unrelated, that is a sporadically appearing bug we haven't been able to pin down yet. I'm merging this.
@jackhansom Thank you!

**Pull request merged**